core/src/commonMain/kotlin/com/saveourtool/processbuilder/ProcessBuilder.kt

```diff
@@ -0,0 +1,211 @@
+        } catch (ex: ProcessTimeoutException) {
+            fs.deleteRecursively(tmpDir)
+            throw ex
+        } catch (ex: Exception) {
+            fs.deleteRecursively(tmpDir)
+            logErrorAndThrowProcessBuilderException(ex.message ?: "Couldn't execute $cmd")
+        }
+
+        val stdout = fs.readLines(stdoutFile)
+        val stderr = fs.readLines(stderrFile)
+
+        fs.myDeleteRecursively(tmpDir)
+        logger.trace { "Removed temp directory $tmpDir" }
+        if (stderr.isNotEmpty()) {
+            logger.debug { "stderr of `$command`:\t${stderr.joinToString("\t")}" }
+        }
+        redirectTo?.let {
+            fs.write(redirectTo) {
+                write(stdout.joinToString("\n").encodeToByteArray())
+                write(stderr.joinToString("\n").encodeToByteArray())
+            }
+        } ?: logger.trace { "Execution output:\t$stdout" }
+        return ExecutionResult(status, stdout, stderr)
+    }
+
+    private fun modifyCmd(
+        command: String,
+        directory: String,
+        processBuilderInternal: ProcessBuilderInternal,
+    ): String {
+        // If we need to step out into some directory before execution
+        val cdCmd = if (directory.isNotBlank()) {
+            if (isCurrentOsWindows()) {
+                "cd /d $directory && "
+            } else {
+                "cd $directory && "
+            }
+        } else {
+            ""
+        }
+        // Additionally process command for Windows, it it contain `echo`
+        val commandWithEcho = cdCmd + if (isCurrentOsWindows()) {
+            processCommandWithEcho(command)
+        } else {
+            command
+        }
+        logger.trace { "Modified cmd: $commandWithEcho" }
+        // Finally, make platform dependent adaptations
+        return processBuilderInternal.prepareCmd(commandWithEcho)
+    }
+
+    /**
+     * Log error message and throw exception
+     *
+     * @param errMsg error message
+     * @throws ProcessExecutionException
+     */
+    private fun logErrorAndThrowProcessBuilderException(errMsg: String): Nothing {
+        logger.error { errMsg }
+        throw ProcessExecutionException(errMsg)
+    }
+
+    companion object {
+        /**
+         * Check whether there are exists `echo` commands, and process them, since in Windows
+         * `echo` adds extra whitespaces and newlines. This method will remove them
+         *
+         * @param command command to process
+         * @return unmodified command, if there is no `echo` subcommands, otherwise add parameter `set /p=` to `echo`
+         */
+        @Suppress("ReturnCount")
+        fun processCommandWithEcho(command: String): String {
+            if (!command.contains("echo")) {
+                return command
+            }
+            // Command already contains correct signature.
+            // We also believe not to met complex cases: `echo a; echo | set /p="a && echo b"`
```

Comment:
Initially, we used PB as executor only for save-cli, which directed to execute only some tests, and it's was okay to not expect to met such cases

But now, this comment seems to be as `TODO` or `issue`, that shouldn't be ignored for this library

Reply:
https://github.com/saveourtool/process-builder-multiplatform/issues/5